### DISCUSSION TITLE
Disabling update of existing packages

### DISCUSSION THREAD

**Discussion by cavo789** [quarto-dev/quarto-cli#7380] [Q&A]
### Description

Hello

I'm (trying) to run Quarto from a Docker container. The Docker image has been built right now and the container is up and running.

I was expecting that everything was up-to-date.  I mean, I just build the Docker image and one of the instruction is `quarto tool install tinytex` so that one is up-to-date

But, when running Quarto using my Docker container, using `quarto render /project/input/readme.md --log /tmp/quarto.log --to pdf`, I see the output below.

Of course, since it's a container, it has no sense to update packages in my container since once the conversion has been made, the container is removed so, on the next call, boum, update packages again and again ...

Did you know how can avoid this ?  I mean : can I "force" Quarto (or another tool) to update every dependencies during the creation of my Docker image or, if not possible, how can I disable packages update ?

Many thanks!!

----

My output :

```text
quarto render /project/input/readme.md --log /tmp/quarto.log --to pdf
pandoc
  to: latex
  output-file: readme.tex
  template: /project/templates/eisvogel/eisvogel.latex
  standalone: true
  pdf-engine: xelatex
  variables:
    graphics: true
    tables: true
  default-image-extension: pdf

running xelatex - 1
  This is XeTeX, Version 3.141592653-2.6-0.999995 (TeX Live 2023) (preloaded format=xelatex)
   restricted \write18 enabled.
  entering extended mode

updating tlmgr

updating existing packages
finding package for fvextra.sty
> 1 package to install
> installing fvextra (1 of 1)
  This is XeTeX, Version 3.141592653-2.6-0.999995 (TeX Live 2023) (preloaded format=xelatex)
   restricted \write18 enabled.
  entering extended mode

finding package for footnotebackref.sty
> 1 package to install
> installing footnotebackref (1 of 1)
  This is XeTeX, Version 3.141592653-2.6-0.999995 (TeX Live 2023) (preloaded format=xelatex)
   restricted \write18 enabled.
  entering extended mode

finding package for pagecolor.sty
> 1 package to install
> installing pagecolor (1 of 1)
  This is XeTeX, Version 3.141592653-2.6-0.999995 (TeX Live 2023) (preloaded format=xelatex)
   restricted \write18 enabled.
  entering extended mode

finding package for sourcesanspro.sty
> 1 package to install
> installing sourcesanspro (1 of 1)
  This is XeTeX, Version 3.141592653-2.6-0.999995 (TeX Live 2023) (preloaded format=xelatex)
   restricted \write18 enabled.
  entering extended mode

finding package for sourcecodepro.sty
> 1 package to install
> installing sourcecodepro (1 of 1)
  This is XeTeX, Version 3.141592653-2.6-0.999995 (TeX Live 2023) (preloaded format=xelatex)
   restricted \write18 enabled.
  entering extended mode

finding package for titling.sty
> 1 package to install
> installing titling (1 of 1)
  This is XeTeX, Version 3.141592653-2.6-0.999995 (TeX Live 2023) (preloaded format=xelatex)
   restricted \write18 enabled.
  entering extended mode

running xelatex - 2
  This is XeTeX, Version 3.141592653-2.6-0.999995 (TeX Live 2023) (preloaded format=xelatex)
   restricted \write18 enabled.
  entering extended mode

Output created: readme.pdf
```

**Comment by cscheid** [comment]
I think this is a tinytex question, not a quarto question.

**Reply by cavo789** [reply]
Thanks for the very quick answer.

You are certainly right, I have searched a lot without finding where to start. Therefore my question here.

Did you have some tips ? URLs ? 

Thanks !

**Reply by mcanouil** [reply]
The TinyTex repository is a good start (and not that hard to find 😉): https://github.com/rstudio/tinytex

Documentation: https://yihui.org/tinytex/

**Comment by cavo789** [comment]
Thanks a lot.  I've still searched during the week-end , discover the existence of commands like `~/.TinyTeX/bin/x86_64-linux/tlmgr update --self --all` and `~/.TinyTeX/bin/x86_64-linux/fmtutil-sys --all` but without effect.

I've then created a ticket on the TinyTex repo (https://github.com/rstudio/tinytex/issues/426)

Have a nice day.

**Comment by cderv** [comment]
I am posting back here the answer I gave at **tinytex** R repo. This question is mainly about using TeX Live in a docker container so that you can render document using Quarto. So this is as much a Quarto question that a **tinytex** question, as both are using TinyTeX distribution based on TeX Live. Also depending on the document you are trying to render, it could be a Quarto related improvment if some of the missing CTAN package in the install you get by `quarto tools install tinytex` are for default `format: pdf` feature. 

Anyhow, here is my reply on other thread for any future readers: 

> I was expecting that everything was up-to-date. I mean, I just build the Docker image and one of the instruction is quarto tool install tinytex so that one is up-to-date

When you install `quarto tool install tinytex`, you are installing one of the bundle from https://github.com/rstudio/tinytex-releases/ 
This bundles is built at a certain date, and include a certain set of package but not all CTAN. For now, Quarto install a Monthly build using the TinyTeX bundle. 

So if your document requires some specific packages from CTAN that are missing from the install, quarto will try to install them as part of the auto-install feature (https://quarto.org/docs/output-formats/pdf-engine.html#quarto-pdf-engine). And updating is part of the process, also because `TeXLive` is updated almost daily ! 

When using docker with LaTeX, you need to deal with installing everything you need once, or allowing updates indeed. 

Several option for you: 

* Install TinyTeX and deactivate auto install from Quarto (https://quarto.org/docs/output-formats/pdf-engine.html#package-installation)
	* Then install manually in your docker container all the packages that are missing and that your document requires. 
	* By deactivating auto install, you will find them by rendering your document and looking at the error. 

  This is usual process for a docker image: You need to install all the requirement you have once at build time. This includes CTAN packages for your LaTeX rendering. 

* Install a newer version of TinyTeX in you docker container. Quarto install function does not allow that yet, but you can install TinyTeX with other ways. However, if you don't build your container often you will have the same issue, and solution above is still needed. 

* Install the `TinyTeX-2` bundle which includes full CTAN 
	> TinyTeX-2 contains the scheme-full scheme of TeX Live, which means all LaTeX packages that you could possibly install from CTAN. This is the largest bundle, and only available in the [daily release](https://github.com/rstudio/tinytex-releases/releases/daily).

  This way you docker image will be pinned to a specific version, with no missing package, meaning no update should be triggered. 

  However, your image will be heavy ! 

## About manually installing package 

See https://tug.org/texlive/doc/tlmgr.html#install-option...-pkg , `tlmgr install <pkgname>` is the command to run  when a CTAN package is missing. 

Though the error message won't give you the package name but the filename missing, which you need to look for it using `tlmgr search --file --global <filename>`

````powershell
> tlmgr search --file --global /fvextra.sty
tlmgr.pl: package repository https://mirror.las.iastate.edu/tex-archive/systems/texlive/tlnet/ (verified)
fvextra:
        texmf-dist/tex/latex/fvextra/fvextra.sty
````

This tells you that `fvextra` is the package to install, which you can do with `tlmgr install fvextra`. 

Once you have the list of all the package you need, you can have one call in your dockerfile `tlmgr install fvextra footnotebackref`

This is only TeX Live knownledge - nothing Quarto specific. 

Though, Quarto do that for you through the auto install - if you want to have those install when building, you need to add them in your install.

**Comment by cavo789** [comment]
Thanks Christophe.  Your help was precious, problem solved.

**Comment by BlackHC** [comment]
Passing `-M latex-auto-install=false` on the command line does the trick!

**Reply by cderv** [reply]
Yes this is the opt-out way: https://quarto.org/docs/output-formats/pdf-engine.html#quarto-pdf-engine

Command flag or through YAML configuration for project or document
